packages/inscription/src/inscription/inscription-ui.tsx

```diff
@@ -55,38 +54,45 @@ export class InscriptionUi extends GLSPAbstractUIExtension implements IActionHan
     return 'inscription-ui-container';
   }
 
+  protected resolveInscriptionClient(client: InscriptionClientJsonRpc) {
+    this.resolvedInscriptionClient = client;
+    return client;
+  }
+
   protected initializeContents(containerElement: HTMLElement) {
+    super.initializeContents(containerElement);
     this.changeUiVisiblitiy(false);
     this.inscriptionContext = this.initInscriptionContext();
     this.queryClient = initQueryClient();
-    this.inscriptionClient = this.startInscriptionClient();
-    this.root = createRoot(containerElement);
+    this.inscriptionClient = this.startInscriptionClient().then(client => this.resolveInscriptionClient(client));
   }
 
-  private updateInscriptionUi() {
-    if (!this.inscriptionElement) {
+  protected render(): React.ReactNode {
+    const element = this.inscriptionElement;
+    if (!element) {
       return;
     }
-    const element = this.inscriptionElement;
-    this.inscriptionClient?.then(client => {
-      this.root.render(
-        <React.StrictMode>
-          <ClientContextProvider client={client}>
-            <QueryProvider client={this.queryClient}>
-              <InscriptionView
-                app={this.inscriptionContext.app}
-                pmv={this.inscriptionContext.pmv}
-                pid={element}
-                outline={{
-                  selection: this.selectionService.getSelectedElementIDs()[0],
-                  onClick: id => this.selectFromOutline(id)
-                }}
-              />
-            </QueryProvider>
-          </ClientContextProvider>
-        </React.StrictMode>
-      );
-    });
+    if (!this.resolvedInscriptionClient) {
+      this.inscriptionClient?.then(() => this.update());
+      return;
+    }
+    return (
+      <React.StrictMode>
```

Comment:
I think this should be part of the react UI extension itself

Reply:
Good point! I added it directly to the React UI extension class.